rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -419,6 +430,26 @@ bool RecorderImpl::is_paused()
   return paused_.load();
 }
 
+void RecorderImpl::stop_discovery()
+{
+  std::lock_guard<std::recursive_mutex> lock(state_transaction_mutex_);
+  if (stop_discovery_ == true) {
```

Comment:
@fujitatomoya I don't see a reason or rationale for protecting discovery start/stop operations with mutex. 
After some analysis, I came to the conclusion it would be enough to have an atomic `stop_discovery_` variable. 
I don't see possibilities for race conditions or undefined behavior in discovery start/stop operations.

If you agree please remove mutex lock here and use regular **non**-recursive mutex for the `state_transition_mutex_`.

Reply:
Hi @fujitatomoya 
For the case you described 
> e.g) 10 threads possibly call on discovery_future_.wait_for that leads to the timeout except one.

I think that protection with ` if (stop_discovery_.exchange(true)) {` would be enough?
i.e. It would look like that without locking the mutex
```cpp
void RecorderImpl::stop_discovery()
{
  if (stop_discovery_.exchange(true)) {
    RCLCPP_DEBUG(
      node->get_logger(),
      "Recorder topic discovery has already been stopped or not running.");
  } else {
    if (discovery_future_.valid()) {
      auto status = discovery_future_.wait_for(2 * record_options_.topic_polling_interval);
      if (status != std::future_status::ready) {
        RCLCPP_ERROR_STREAM(
          node->get_logger(),
          "discovery_future_.wait_for(" << record_options_.topic_polling_interval.count() <<
            ") return status: " <<
            (status == std::future_status::timeout ? "timeout" : "deferred"));
      }
    }
  }
}
```
For other cases when we are calling `stop_discovery()` from inner class members (RecorderImpl::stop()) it is already protected with mutex.

Reply:
If considering other cases when we can potentially change `stop_discovery_` atomic variable.
We are changing it only in `RecorderImpl::record()` and `RecorderImpl::stop()` or from destructor when calling `RecorderImpl::stop()` in all that scenarious we aleady locking the transitional mutex on the upper layer.

Reply:
Only possible situation for race condition could be if we would call concurrently from different threads `RecorderImpl::record()` and `RecorderImpl::stop_discovery()`.
In this case, it will be better to create a separate mutex for discovery and wrap start discovery routines to a separate method e.g. `Recorder::start_discovery()`. 
I think it might be useful in the future to link those API to the service calls.

Reply:
@MichaelOrlov thanks!

`stop_discovery_.exchange(true)` should do work!